processing/src/main/java/io/druid/query/groupby/epinephelinae/ConcurrentGrouper.java

```diff
@@ -216,15 +230,53 @@ public void reset()
       throw new ISE("Grouper is closed");
     }
 
-    final List<Iterator<Entry<KeyType>>> iterators = new ArrayList<>(groupers.size());
+    return Groupers.mergeIterators(
+        sorted && isParallelSortAvailable() ? parallelSortAndGetGroupersIterator() : getGroupersIterator(sorted),
+        sorted ? keyObjComparator : null
+    );
+  }
 
-    for (Grouper<KeyType> grouper : groupers) {
-      synchronized (grouper) {
-        iterators.add(grouper.iterator(sorted));
-      }
+  private boolean isParallelSortAvailable()
+  {
+    return grouperSorter != null && concurrencyHint > 1;
+  }
+
+  private List<Iterator<Entry<KeyType>>> parallelSortAndGetGroupersIterator()
+  {
+    // The number of groupers is same with the number of processing threads in grouperSorter
+    final List<ListenableFuture<Iterator<Entry<KeyType>>>> futures = groupers
+        .stream()
+        .map(grouper ->
+                 grouperSorter.submit(
+                     new AbstractPrioritizedCallable<Iterator<Entry<KeyType>>>(priority)
```

Comment:
Maybe turn this into non-abstract class that accepts two constructor parameters: priority and lambda. Looks like it should simplify most use sites.